File: frontend/src/services/debug.ts
/**
 * Debug API Service
 *
 * Provides methods for the Phase 1 AI parse analysis debug tool.
 */

const API_BASE = '/api/debug';

// Types matching backend schemas
export interface ParsedIntent {
  request_type: string;
  target_name: string | null;
  keywords_found: string[];
  parse_notes: string;
  reasoning: string;
  list_position: number;
  needs_clarification: boolean;
  temporal_info: {
    date: string | null;
    is_superseded: boolean;
    supersedes_reason: string | null;
  } | null;
}

export interface ParseAnalysisItem {
  id: string;
  original_request_id: string;
  requester_name: string | null;
  requester_cm_id: number | null;
  source_field: string | null;
  original_text: string | null;
  parsed_intents: ParsedIntent[];
  is_valid: boolean;
  error_message: string | null;
  token_count: number | null;
  processing_time_ms: number | null;
  prompt_version: string | null;
  created: string | null;
}

export interface ParseAnalysisDetailItem extends ParseAnalysisItem {
  ai_raw_response: Record<string, unknown> | null;
}

export interface ParseAnalysisListResponse {
  items: ParseAnalysisItem[];
  total: number;
}

export interface OriginalRequestItem {
  id: string;
  requester_name: string | null;
  requester_cm_id: number | null;
  source_field: string;
  original_text: string;
  year: number;
  processed: boolean;
}

export interface OriginalRequestsListResponse {
  items: OriginalRequestItem[];
  total: number;
}

export interface Phase1OnlyRequest {
  original_request_ids: string[];
  force_reparse: boolean;
}

export interface Phase1OnlyResponse {
  results: ParseAnalysisItem[];
  total_tokens: number;
}

export interface ClearAnalysisResponse {
  deleted_count: number;
}

export type SourceFieldType = 'bunk_with' | 'not_bunk_with' | 'bunking_notes' | 'internal_notes';

export interface ParseAnalysisFilters {
  session_cm_id?: number | undefined;
  source_field?: SourceFieldType | undefined;
  limit?: number | undefined;
  offset?: number | undefined;
}

export interface OriginalRequestsFilters {
  year: number;
  session_cm_id?: number;
  source_field?: SourceFieldType;
  limit?: number;
}

// New types for fallback pattern
export type ParseResultSource = 'debug' | 'production' | 'none';

export interface OriginalRequestWithStatus {
  id: string;
  requester_name: string | null;
  requester_cm_id: number | null;
  source_field: string;
  original_text: string;
  year: number;
  has_debug_result: boolean;
  has_production_result: boolean;
}

export interface OriginalRequestsWithParseResponse {
  items: OriginalRequestWithStatus[];
  total: number;
}

export interface ParseResultWithSource {
  source: ParseResultSource;
  id: string | null;
  original_request_id: string | null;
  requester_name: string | null;
  requester_cm_id: number | null;
  source_field: string | null;
  original_text: string | null;
  parsed_intents: ParsedIntent[];
  is_valid: boolean;
  error_message: string | null;
  token_count: number | null;
  processing_time_ms: number | null;
  prompt_version: string | null;
  created: string | null;
}

export interface OriginalRequestsWithStatusFilters {
  year: number;
  session_cm_ids?: number[] | undefined;
  source_field?: SourceFieldType | undefined;
  limit?: number | undefined;
  offset?: number | undefined;
}

// Types for grouped-by-camper view
export interface FieldParseResult {
  original_request_id: string;
  source_field: string;
  original_text: string;
  has_debug_result: boolean;
  has_production_result: boolean;
}

export interface CamperGroupedRequests {
  requester_cm_id: number;
  requester_name: string;
  fields: FieldParseResult[];
}

export interface GroupedRequestsResponse {
  items: CamperGroupedRequests[];
  total: number;
}

export interface GroupedRequestsFilters {
  year: number;
  session_cm_ids?: number[] | undefined;
  source_field?: SourceFieldType | undefined;
  limit?: number | undefined;
}

export interface ScopedClearFilters {
  session_cm_ids?: number[] | undefined;
  source_field?: SourceFieldType | undefined;
}

// Prompt Editor Types

export interface PromptListItem {
  name: string;
  filename: string;
  modified_at: string | null;
}

export interface PromptListResponse {
  prompts: PromptListItem[];
}

export interface PromptContentResponse {
  name: string;
  content: string;
  modified_at: string | null;
}

export interface PromptUpdateRequest {
  content: string;
}

export interface PromptUpdateResponse {
  name: string;
  success: boolean;
}

export const debugService = {
  /**
   * List parse analysis results with optional filters
   */
  async listParseAnalysis(
    filters: ParseAnalysisFilters,
    fetchWithAuth: (url: string, options?: RequestInit) => Promise<Response>
  ): Promise<ParseAnalysisListResponse> {
    const params = new URLSearchParams();
    if (filters.session_cm_id) params.set('session_cm_id', String(filters.session_cm_id));
    if (filters.source_field) params.set('source_field', filters.source_field);
    if (filters.limit) params.set('limit', String(filters.limit));
    if (filters.offset) params.set('offset', String(filters.offset));

    const url = `${API_BASE}/parse-analysis${params.toString() ? `?${params}` : ''}`;
    const response = await fetchWithAuth(url);

    if (!response.ok) {
      throw new Error('Failed to fetch parse analysis results');
    }
    return response.json();
  },

  /**
   * Get detailed parse analysis result by ID
   */
  async getParseAnalysisDetail(
    id: string,
    fetchWithAuth: (url: string, options?: RequestInit) => Promise<Response>
  ): Promise<ParseAnalysisDetailItem> {
    const response = await fetchWithAuth(`${API_BASE}/parse-analysis/${id}`);

    if (!response.ok) {
      throw new Error('Failed to fetch parse analysis detail');
    }
    return response.json();
  },

  /**
   * Run Phase 1 parsing on selected original requests
   */
  async parsePhase1Only(
    request: Phase1OnlyRequest,
    fetchWithAuth: (url: string, options?: RequestInit) => Promise<Response>
  ): Promise<Phase1OnlyResponse> {
    const response = await fetchWithAuth(`${API_BASE}/parse-phase1-only`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new Error(error.detail || 'Failed to run Phase 1 parsing');
    }
    return response.json();
  },

  /**
   * Clear parse analysis results (with optional scoped filters)
   */
  async clearParseAnalysis(
    fetchWithAuth: (url: string, options?: RequestInit) => Promise<Response>,
    filters?: ScopedClearFilters
  ): Promise<ClearAnalysisResponse> {
    const params = new URLSearchParams();
    if (filters?.session_cm_ids) {
      filters.session_cm_ids.forEach((cmId) => {
        params.append('session_cm_id', String(cmId));
      });
    }
    if (filters?.source_field) params.set('source_field', filters.source_field);

    const url = `${API_BASE}/parse-analysis${params.toString() ? `?${params}` : ''}`;
    const response = await fetchWithAuth(url, {
      method: 'DELETE',
    });

    if (!response.ok) {
      throw new Error('Failed to clear parse analysis results');
    }
    return response.json();
  },

  /**
   * Clear a single parse analysis result by original request ID
   */
  async clearSingleParseAnalysis(
    originalRequestId: string,
    fetchWithAuth: (url: string, options?: RequestInit) => Promise<Response>
  ): Promise<ClearAnalysisResponse> {
    const response = await fetchWithAuth(
      `${API_BASE}/parse-analysis/by-original/${encodeURIComponent(originalRequestId)}`,
      { method: 'DELETE' }
    );

    if (!response.ok) {
      throw new Error('Failed to clear parse analysis result');
    }
    return response.json();
  },

  /**
   * List original bunk requests for debug selection
   */
  async listOriginalRequests(
    filters: OriginalRequestsFilters,
    fetchWithAuth: (url: string, options?: RequestInit) => Promise<Response>
  ): Promise<OriginalRequestsListResponse> {
    const params = new URLSearchParams();
    params.set('year', String(filters.year));
    if (filters.session_cm_id) params.set('session_cm_id', String(filters.session_cm_id));
    if (filters.source_field) params.set('source_field', filters.source_field);
    if (filters.limit) params.set('limit', String(filters.limit));

    const response = await fetchWithAuth(`${API_BASE}/original-requests?${params}`);

    if (!response.ok) {
      throw new Error('Failed to fetch original requests');
    }
    return response.json();
  },

  /**
   * List original requests with parse status flags (debug/production)
   */
  async listOriginalRequestsWithStatus(
    filters: OriginalRequestsWithStatusFilters,
    fetchWithAuth: (url: string, options?: RequestInit) => Promise<Response>
  ): Promise<OriginalRequestsWithParseResponse> {
    const params = new URLSearchParams();
    params.set('year', String(filters.year));
    // Pass multiple session_cm_id params for array (FastAPI handles repeated query params)
    if (filters.session_cm_ids) {
      filters.session_cm_ids.forEach((cmId) => {
        params.append('session_cm_id', String(cmId));
      });
    }
    if (filters.source_field) params.set('source_field', filters.source_field);
    if (filters.limit) params.set('limit', String(filters.limit));
    if (filters.offset) params.set('offset', String(filters.offset));

    const response = await fetchWithAuth(`${API_BASE}/original-requests-with-parse-status?${params}`);

    if (!response.ok) {
      throw new Error('Failed to fetch original requests with status');
    }
    return response.json();
  },

  /**
   * Get parse result with fallback (debug -> production -> none)
   */
  async getParseResultWithFallback(
    originalRequestId: string,
    fetchWithAuth: (url: string, options?: RequestInit) => Promise<Response>
  ): Promise<ParseResultWithSource> {
    const response = await fetchWithAuth(`${API_BASE}/parse-result/${encodeURIComponent(originalRequestId)}`);

    if (!response.ok) {
      throw new Error('Failed to fetch parse result');
    }
    return response.json();
  },

  /**
   * Get parse results for multiple original requests in one batch call.
   * Much faster than calling getParseResultWithFallback multiple times.
   */
  async getParseResultsBatch(
    originalRequestIds: string[],
    fetchWithAuth: (url: string, options?: RequestInit) => Promise<Response>
  ): Promise<ParseResultWithSource[]> {
    if (originalRequestIds.length === 0) {
      return [];
    }

    const response = await fetchWithAuth(`${API_BASE}/parse-results-batch`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(originalRequestIds),
    });

    if (!response.ok) {
      throw new Error('Failed to fetch parse results batch');
    }
    return response.json();
  },

  /**
   * List original requests grouped by camper (excludes socialize_with)
   */
  async listGroupedRequests(
    filters: GroupedRequestsFilters,
    fetchWithAuth: (url: string, options?: RequestInit) => Promise<Response>
  ): Promise<GroupedRequestsResponse> {
    const params = new URLSearchParams();
    params.set('year', String(filters.year));
    if (filters.session_cm_ids) {
      filters.session_cm_ids.forEach((cmId) => {
        params.append('session_cm_id', String(cmId));
      });
    }
    if (filters.source_field) params.set('source_field', filters.source_field);
    if (filters.limit) params.set('limit', String(filters.limit));

    const response = await fetchWithAuth(`${API_BASE}/original-requests-grouped?${params}`);

    if (!response.ok) {
      throw new Error('Failed to fetch grouped requests');
    }
    return response.json();
  },

  // ============================================================================
  // Prompt Editor Methods
  // ============================================================================

  /**
   * List available prompt files
   */
  async listPrompts(
    fetchWithAuth: (url: string, options?: RequestInit) => Promise<Response>
  ): Promise<PromptListResponse> {
    const response = await fetchWithAuth(`${API_BASE}/prompts`);

    if (!response.ok) {
      throw new Error('Failed to fetch prompts list');
    }
    return response.json();
  },

  /**
   * Get the content of a specific prompt
   */
  async getPrompt(
    name: string,
    fetchWithAuth: (url: string, options?: RequestInit) => Promise<Response>
  ): Promise<PromptContentResponse> {
    const response = await fetchWithAuth(`${API_BASE}/prompts/${encodeURIComponent(name)}`);

    if (!response.ok) {
      if (response.status === 404) {
        throw new Error(`Prompt '${name}' not found`);
      }
      throw new Error('Failed to fetch prompt content');
    }
    return response.json();
  },

  /**
   * Update a prompt's content
   */
  async updatePrompt(
    name: string,
    content: string,
    fetchWithAuth: (url: string, options?: RequestInit) => Promise<Response>
  ): Promise<PromptUpdateResponse> {
    const response = await fetchWithAuth(`${API_BASE}/prompts/${encodeURIComponent(name)}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ content } satisfies PromptUpdateRequest),
    });

    if (!response.ok) {
      if (response.status === 404) {
        throw new Error(`Prompt '${name}' not found`);
      }
      const error = await response.json().catch(() => ({}));
      throw new Error(error.detail || 'Failed to update prompt');
    }
    return response.json();
  },
};

File: api/schemas/debug.py
"""Debug API Schemas

Pydantic models for the debug parse analysis endpoints.
"""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

# Source field enum for filtering
SourceFieldType = Literal["bunk_with", "not_bunk_with", "bunking_notes", "internal_notes"]


class ParseAnalysisFilter(BaseModel):
    """Filter parameters for listing parse analysis results."""

    session_cm_id: int | None = Field(default=None, description="Filter by session CampMinder ID")
    source_field: SourceFieldType | None = Field(default=None, description="Filter by source field type")
    year: int | None = Field(default=None, description="Filter by year")
    limit: int = Field(default=50, ge=1, le=500, description="Maximum number of results")
    offset: int = Field(default=0, ge=0, description="Offset for pagination")


class ParsedIntent(BaseModel):
    """A single parsed intent from Phase 1."""

    request_type: str = Field(description="Type of request (bunk_with, not_bunk_with, etc.)")
    target_name: str | None = Field(description="Extracted target camper name")
    keywords_found: list[str] = Field(default_factory=list, description="Keywords that triggered this intent")
    parse_notes: str = Field(default="", description="Notes about the parse")
    reasoning: str = Field(default="", description="AI reasoning for this parse")
    list_position: int = Field(description="Position in the original text (0-based)")
    needs_clarification: bool = Field(default=False, description="Whether this intent needs clarification")
    temporal_info: dict[str, Any] | None = Field(default=None, description="Temporal conflict information if present")


class ParseAnalysisItem(BaseModel):
    """A single parse analysis result for display."""

    id: str = Field(description="Debug result record ID")
    original_request_id: str = Field(description="Original bunk request record ID")
    requester_name: str | None = Field(description="Name of the camper who made the request")
    requester_cm_id: int | None = Field(description="CampMinder ID of the requester")
    source_field: str | None = Field(description="Source field (bunk_with, bunking_notes, etc.)")
    original_text: str | None = Field(description="Original request text from CSV")
    parsed_intents: list[ParsedIntent] = Field(default_factory=list, description="Parsed intents from Phase 1")
    is_valid: bool = Field(description="Whether parsing succeeded")
    error_message: str | None = Field(default=None, description="Error message if parsing failed")
    token_count: int | None = Field(default=None, description="Number of tokens used")
    processing_time_ms: int | None = Field(default=None, description="Processing time in milliseconds")
    prompt_version: str | None = Field(default=None, description="Version of prompt used")
    created: datetime | None = Field(default=None, description="When the result was created")


class ParseAnalysisDetailItem(ParseAnalysisItem):
    """Extended parse analysis with raw AI response for detail view."""

    ai_raw_response: dict[str, Any] | None = Field(default=None, description="Raw AI response for debugging")


class ParseAnalysisListResponse(BaseModel):
    """Response for listing parse analysis results."""

    items: list[ParseAnalysisItem] = Field(description="List of parse analysis results")
    total: int = Field(description="Total number of results matching filter")


class Phase1OnlyRequest(BaseModel):
    """Request body for running Phase 1 parsing."""

    original_request_ids: list[str] = Field(
        min_length=1,
        description="List of original_bunk_requests record IDs to parse",
    )
    force_reparse: bool = Field(
        default=False,
        description="If true, ignore cache and reparse even if results exist",
    )


class Phase1OnlyResponse(BaseModel):
    """Response from Phase 1 parsing."""

    results: list[ParseAnalysisItem] = Field(description="Parse results for each request")
    total_tokens: int = Field(description="Total tokens used across all parses")


class ClearAnalysisResponse(BaseModel):
    """Response from clearing analysis results."""

    deleted_count: int = Field(description="Number of records deleted")


class OriginalRequestItem(BaseModel):
    """An original bunk request for the debug list."""

    id: str = Field(description="PocketBase record ID")
    requester_name: str | None = Field(description="Name of the camper")
    requester_cm_id: int | None = Field(description="CampMinder ID of the requester")
    source_field: str = Field(description="Source field type")
    original_text: str = Field(description="Original request text")
    year: int = Field(description="Camp year")
    processed: bool = Field(description="Whether this request has been processed")


class OriginalRequestsListResponse(BaseModel):
    """Response for listing original requests."""

    items: list[OriginalRequestItem] = Field(description="List of original requests")
    total: int = Field(description="Total count")


class OriginalRequestWithStatus(BaseModel):
    """An original bunk request with parse status flags."""

    id: str = Field(description="PocketBase record ID")
    requester_name: str | None = Field(description="Name of the camper")
    requester_cm_id: int | None = Field(description="CampMinder ID of the requester")
    source_field: str = Field(description="Source field type")
    original_text: str = Field(description="Original request text")
    year: int = Field(description="Camp year")
    has_debug_result: bool = Field(description="Whether a debug parse result exists")
    has_production_result: bool = Field(description="Whether production bunk_requests exist")


class OriginalRequestsWithParseResponse(BaseModel):
    """Response for listing original requests with parse status."""

    items: list[OriginalRequestWithStatus] = Field(description="List of original requests with status")
    total: int = Field(description="Total count")


# Source type for parse results
ParseResultSource = Literal["debug", "production", "none"]


class ParseResultWithSource(BaseModel):
    """Parse result with source indicator for fallback display."""

    source: ParseResultSource = Field(description="Source of the parse result")
    id: str | None = Field(default=None, description="Debug result record ID (if source is debug)")
    original_request_id: str | None = Field(default=None, description="Original bunk request record ID")
    requester_name: str | None = Field(default=None, description="Name of the camper")
    requester_cm_id: int | None = Field(default=None, description="CampMinder ID of the requester")
    source_field: str | None = Field(default=None, description="Source field type")
    original_text: str | None = Field(default=None, description="Original request text")
    parsed_intents: list[ParsedIntent] = Field(default_factory=list, description="Parsed intents")
    is_valid: bool = Field(default=True, description="Whether parsing succeeded")
    error_message: str | None = Field(default=None, description="Error message if parsing failed")
    token_count: int | None = Field(default=None, description="Number of tokens used")
    processing_time_ms: int | None = Field(default=None, description="Processing time in milliseconds")
    prompt_version: str | None = Field(default=None, description="Version of prompt used")
    created: datetime | None = Field(default=None, description="When the result was created")


# Prompt Editor Schemas


class PromptListItem(BaseModel):
    """A single prompt file in the list."""

    name: str = Field(description="Prompt name (without extension)")
    filename: str = Field(description="Full filename")
    modified_at: datetime | None = Field(default=None, description="Last modification time")


class PromptListResponse(BaseModel):
    """Response for listing available prompts."""

    prompts: list[PromptListItem] = Field(description="List of available prompts")


class PromptContentResponse(BaseModel):
    """Response for getting prompt content."""

    name: str = Field(description="Prompt name")
    content: str = Field(description="Prompt content")
    modified_at: datetime | None = Field(default=None, description="Last modification time")


class PromptUpdateRequest(BaseModel):
    """Request body for updating a prompt."""

    content: str = Field(min_length=1, description="New prompt content")


class PromptUpdateResponse(BaseModel):
    """Response for updating a prompt."""

    name: str = Field(description="Prompt name")
    success: bool = Field(description="Whether update succeeded")


# Grouped by Camper Schemas (Phase 3)


class FieldParseResult(BaseModel):
    """A single field's parse result within a camper group."""

    original_request_id: str = Field(description="Original bunk request record ID")
    source_field: str = Field(description="Source field type")
    original_text: str = Field(description="Original request text")
    has_debug_result: bool = Field(description="Whether debug parse result exists")
    has_production_result: bool = Field(description="Whether production bunk_requests exist")


class CamperGroupedRequests(BaseModel):
    """Requests grouped by camper with their field parse results."""

    requester_cm_id: int = Field(description="CampMinder ID of the requester")
    requester_name: str = Field(description="Name of the camper")
    fields: list[FieldParseResult] = Field(description="List of fields for this camper")


class GroupedRequestsResponse(BaseModel):
    """Response for grouped original requests endpoint."""

    items: list[CamperGroupedRequests] = Field(description="List of camper groups")
    total: int = Field(description="Total number of campers")
